pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionAssignmentTailer.java

```diff
@@ -116,4 +147,51 @@ public void processAssignment(Message<byte[]> msg) {
             this.functionRuntimeManager.processAssignment(assignment);
         }
     }
+    
+    private Reader<byte[]> createReader() throws PulsarClientException {
+        MessageId startMessageId = lastMessageId == null ? MessageId.earliest : lastMessageId;
+        log.info("Assignment tailer will start reading from message id {}", startMessageId);
+
+        return readerBuilder
+                .subscriptionRolePrefix(workerConfig.getWorkerId() + "-function-assignment-tailer")
+                .readerName(workerConfig.getWorkerId() + "-function-assignment-tailer")
+                .topic(workerConfig.getFunctionAssignmentTopic())
+                .readCompacted(true)
+                .startMessageId(startMessageId)
+                .create();
+    }
+
+    private Thread getTailerThread() {
+        Thread t = new Thread(() -> {
+            while (isRunning) {
+                try {
+                    Message<byte[]> msg = reader.readNext(5, TimeUnit.SECONDS);
+                    if (msg == null) {
+                        if (exitOnEndOfTopic && !reader.hasMessageAvailable()) {
+                            break;
+                        }
+                    } else {
+                        processAssignment(msg);
+                        // keep track of the last message read
+                        lastMessageId = msg.getMessageId();
+                    }
+                } catch (Throwable th) {
+                    if (isRunning) {
```

Comment:
should we check for exitOnEndofTopic as well?

Reply:
I don't think we need to since even if "exitOnEndOfTopic" is set ,"isRunning" will still be set to true and any error will be bubbled up as expected